### PR TITLE
confdb: fix unused branch checks and trim indexes

### PR DESCRIPTION
**Pull request by miguelpires**
We check that values passed into Set are completely used by the paths of rules the write matched to. If they're not, we reject the write to prevent people from thinking they're writing something which turns out to be discarded. The check was not taking into account previous paths
covering a particular branch of the value which this commit fixes.
This also trims list indexes so we don't store values as "001". I considered parsing the value into an int but it just adds more code converting it back into strings in various places so I opted against it.

### CODE REVIEW

File: confdb/confdb_test.go [hunk 1]
```diff
@@ -2744,6 +2744,27 @@ func (s *viewSuite) TestViewSetErrorIfValueContainsUnusedParts(c *C) {
 	}
 }
 
+func (s *viewSuite) TestSetPathChecksValueCoveredByMultipleBranches(c *C) {
```
Comment (andrewphelpsj) [inline confdb/confdb_test.go:2747]: This can still pass on master due to random map iteration order:
```
phelps@arch ~/s/m/confdb> go test -check.f TestSetPathChecksValueCoveredByMultipleBranches

----------------------------------------------------------------------
FAIL: confdb_test.go:5225: viewSuite.TestSetPathChecksValueCoveredByMultipleBranches

confdb_test.go:5243:
    c.Assert(err, IsNil)
... value *confdb.BadRequestError = &confdb.BadRequestError{viewID:"acc/confdb/foo", operation:"set", request:"foo", cause:"internal error: cannot use unmatched part \"bar\" as key in map[other:value]"} ("cannot set \"foo\" through confdb view acc/confdb/foo: internal error: cannot use unmatched part \"bar\" as key in map[other:value]")

OOPS: 0 passed, 1 FAILED
--- FAIL: Test (0.01s)
FAIL
exit status 1
FAIL	github.com/snapcore/snapd/confdb	0.017s
phelps@arch ~/s/m/confdb> go test -check.f TestSetPathChecksValueCoveredByMultipleBranches
OK: 1 passed
PASS
ok  	github.com/snapcore/snapd/confdb	0.009s
```

**Review by andrewphelpsj — COMMENTED** [review]
Thanks! One thing about the test

File: confdb/confdb.go [hunk 2]
```diff
@@ -843,7 +843,11 @@ func ParsePathIntoAccessors(path string, opts ParseOptions) ([]Accessor, error)
 			if opts.ForbidIndexes {
 				return nil, fmt.Errorf("invalid subkey %q: view paths cannot have literal indexes (only index placeholders)", subkey)
 			}
-			accessors = append(accessors, newIndex(subkey[1:len(subkey)-1], part.filters))
+			index := strings.TrimLeft(subkey[1:len(subkey)-1], "0")
```
Comment (andrewphelpsj) [inline confdb/confdb.go:846]: This illuminated an odd edge case in the current implementation, even before this branch:
```go
// this test that sets 10 elements in the list passes
func (*viewSuite) TestSetListIndexesInNumericOrderSingleDigit(c *C) {
	schema, err := confdb.NewSchema("acc", "confdb", map[string]any{
		"foo": map[string]any{
			"rules": []any{
				map[string]any{
					"request": "settings[{n}]",
					"storage": "items[{n}]",
				},
			},
		},
	}, confdb.NewJSONSchema())
	c.Assert(err, IsNil)

	// expanding ten elements produces writes to items[0] through items[9].
	// string sorting matches numeric order for these single-digit indexes,
	// so each write appends the next element without leaving a gap.
	value := []any{
		"A", "B", "C", "D", "E", "F", "G", "H", "I", "J",
	}

	bag := confdb.NewJSONDatabag()
	err = schema.View("foo").Set(bag, "settings", value)
	c.Assert(err, IsNil)
}

// this test that sets 11 elements in the list fails
func (*viewSuite) TestSetListIndexesInNumericOrder(c *C) {
	schema, err := confdb.NewSchema("acc", "confdb", map[string]any{
		"foo": map[string]any{
			"rules": []any{
				map[string]any{
					"request": "settings[{n}]",
					"storage": "items[{n}]",
				},
			},
		},
	}, confdb.NewJSONSchema())
	c.Assert(err, IsNil)

	// expanding eleven elements produces a write to items[10]. string sorting
	// incorrectly places it before items[1], attempting to leave a gap when the
	// list has only one element.
	value := []any{
		"A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K",
	}

	bag := confdb.NewJSONDatabag()
	err = schema.View("foo").Set(bag, "settings", value)

	// fails with: cannot access "items[10]": list has length 1
	c.Assert(err, IsNil)
}
```

That second test fails somewhat unexpectedly, since we end up sorting by accessor:
```go
// sort again since we may have unpacked a list into many expanded matches.
// Since list Set()s depend on the length of the existing list, the order matters
getAccs = func(i int) []Accessor { return expandedMatches[i].storagePath }
sort.Slice(expandedMatches, byAccessor(getAccs))
```

I think, given the comment there, we want to be able to support this? It isn't super clear to me from the existing tests.

**Review by andrewphelpsj — COMMENTED** [review]
Some unit tests that I think show a bug. Maybe `byAccessor` needs to consider the `AccessorType` of each `Accessor`?

File: confdb/confdb.go (hunk 2 above)
Comment (miguelpires) [reply confdb/confdb.go:846]: Good catch, I've fixed that now

File: confdb/confdb_test.go (hunk 1 above)
Comment (miguelpires) [reply confdb/confdb_test.go:2747]: I noticed that but I think any test for this will need a path that includes another and going over the more specific before the more generic will make this pass anyway. I didn't think it was worth it to sort them to a deterministic order since the bug was being fixed but it's easily done

File: confdb/confdb_test.go [hunk 3]
```diff
@@ -2744,6 +2744,26 @@ func (s *viewSuite) TestViewSetErrorIfValueContainsUnusedParts(c *C) {
...
+	schema, err := confdb.NewSchema("acc", "confdb", map[string]any{
+		"foo": map[string]any{
+			"rules": []any{
+				map[string]any{"request": "foo.bar.baz", "storage": "foo.bar.baz"},
```
Comment (maykathm) [inline confdb/confdb_test.go:2752]: Since you added the sorting, this should no longer be non-deterministic, no? So we can add this back and cover your fix.
```suggestion
				map[string]any{"request": "foo.bar", "storage": "foo.bar"},
				map[string]any{"request": "foo.bar.baz", "storage": "foo.bar.baz"},
```

Comment (miguelpires) [reply confdb/confdb_test.go:2752]: yup, that's a mistake, it was an earlier commit before the sorting which I meant to drop instead of squash. Fixed now, thank you

**Review by maykathm — APPROVED** [review]
LGTM

**Review by andrewphelpsj — APPROVED** [review]
Thanks!
